openhands/runtime/utils/runtime_build.py

```diff
@@ -224,160 +117,181 @@ def build_runtime_image(
-    """Build and tag the sandbox image.
-    The image will be tagged as both:
-        - target_image_hash_tag
-        - target_image_tag
+def prep_build_folder(
+    build_folder: Path,
+    base_image: str,
+    build_from_scratch: bool,
+    extra_deps: str | None,
+):
+    # Copy the source code to directory. It will end up in build_folder/code
+    # If package is not found, build from source code
+    openhands_source_dir = Path(openhands.__file__).parent
+    project_root = openhands_source_dir.parent
+    logger.info(f'Building source distribution using project root: {project_root}')
+
+    # Copy the 'openhands' directory (Source code)
+    shutil.copytree(
+        openhands_source_dir,
+        Path(build_folder, 'code', 'openhands'),
+        ignore=shutil.ignore_patterns(
+            '.*/',
+            '__pycache__/',
+            '*.pyc',
+        ),
+    )
 
-    Parameters:
-    - docker_folder (str): the path to the docker build folder
-    - runtime_builder (RuntimeBuilder): the runtime builder instance
-    - target_image_repo (str): the repository name for the target image
-    - target_image_hash_tag (str): the *hash* tag for the target image that is calculated based
-        on the contents of the docker build folder (source code and Dockerfile)
-        e.g. 1234567890abcdef
-    -target_image_tag (str): the tag for the target image that's generic and based on the base image name
-        e.g. oh_v0.9.3_image_ubuntu_tag_22.04
-    """
-    target_image_hash_name = f'{target_image_repo}:{target_image_hash_tag}'
-    target_image_generic_name = f'{target_image_repo}:{target_image_tag}'
+    # Copy pyproject.toml and poetry.lock files
+    for file in ['pyproject.toml', 'poetry.lock']:
+        src = Path(openhands_source_dir, file)
+        if not src.exists():
+            src = Path(project_root, file)
+        shutil.copy2(src, Path(build_folder, 'code', file))
 
-    tags_to_add = [target_image_hash_name]
+    # Create a Dockerfile and write it to build_folder
+    dockerfile_content = _generate_dockerfile(
+        base_image,
+        build_from_scratch=build_from_scratch,
+        extra_deps=extra_deps,
+    )
+    with open(Path(build_folder, 'Dockerfile'), 'w') as file:  # type: ignore
+        file.write(dockerfile_content)  # type: ignore
+
+
+def get_hash_for_lock_files(base_image: str):
+    openhands_source_dir = Path(openhands.__file__).parent
+    md5 = hashlib.md5()
+    md5.update(base_image.encode())
+    for file in ['pyproject.toml', 'poetry.lock']:
+        src = Path(openhands_source_dir, file)
+        if not src.exists():
+            src = Path(openhands_source_dir.parent, file)
+        with open(src, 'rb') as f:
+            for chunk in iter(lambda: f.read(4096), b''):
+                md5.update(chunk)
+    # We get away with truncation because we want something that is unique
+    # rather than something that is cryptographically secure
+    result = md5.hexdigest()[:16]
+    return result
```

Comment:
We have max 128 character for docker tag irrc - so maybe we can use a longer lock hash, ideally part of the base image name (so the tag is kinda readable?) - maybe the last 20 char of the image name +  12 char hash?

I think unique while try to be cryptographically secure as possible is not that bad (i want to minimize the chance of spending hours to debug this in the future :( ) - specially we may build with a bunch of different base images in evaluation, which might increase the risk of collision

Reply:
I chose 16 digits not for the sake of being cryptographically secure (because it's not!), but because it gives us a collision space of 1:18,446,744,073,709,551,616 - so you'd win the mega millions lottery twice in a row before this would happen.

 I also considered using base 32 for the hashes, but that only reduces the full hash length from 32 to 25 chars, and I didn't want to increase the complexity - though we could always do this if you are concerned.